hooks/open-telemetry/pkg/otel.go

```diff
@@ -0,0 +1,161 @@
+	h.wg.Add(1)
+	ctx, span := h.tracerClient.tracer().Start(h.ctx, key)
+	ctx, cancel := context.WithCancel(ctx)
+	span.SetAttributes(
+		attribute.String(AttributeFlagKey, hookContext.FlagKey()),
+		attribute.String(AttributeProviderName, hookContext.ProviderMetadata().Name),
+	)
+	h.spans[key].ss = &storedSpan{
+		cancel: cancel,
+		span:   span,
+	}
+	// this goroutine cleans up the span, if the associated context is closed then the stored
+	// span data is removed and the resource is unlocked. This context close can come from either the cancel() method
+	// or from the closing of the parent context outside the scope of the hook
+	go func() {
+		<-ctx.Done()
+		h.spans[key].ss = nil
+		h.spans[key].mu.Unlock()
+		h.wg.Done()
+	}()
+	return nil, nil
+}
+
+// After sets the EvaluatedVariant and EvaluatedValue on the evaluation specific span
+func (h *Hook) After(hookContext of.HookContext, flagEvaluationDetails of.EvaluationDetails, hookHints of.HookHints) error {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if !ok {
+		return errors.New("no span stored for provided hook context")
+	}
+	attributes := []attribute.KeyValue{
+		attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
+	}
+	fmt.Println(flagEvaluationDetails)
+	switch flagEvaluationDetails.FlagType {
+	case of.Boolean:
+		attributes = append(attributes, attribute.Bool(AttributeEvaluatedValue, flagEvaluationDetails.Value.(bool)))
+	case of.String:
+		attributes = append(attributes, attribute.String(AttributeEvaluatedValue, flagEvaluationDetails.Value.(string)))
+	case of.Float:
+		attributes = append(attributes, attribute.Float64(AttributeEvaluatedValue, flagEvaluationDetails.Value.(float64)))
+	case of.Int:
+		attributes = append(attributes, attribute.Int64(AttributeEvaluatedValue, flagEvaluationDetails.Value.(int64)))
+	case of.Object:
+		val, err := json.Marshal(flagEvaluationDetails.Value)
+		if err != nil {
+			return err
+		}
+		attributes = append(attributes, attribute.String(AttributeEvaluatedValue, string(val)))
+	default:
+		return fmt.Errorf("unknown data type received: %d", flagEvaluationDetails.FlagType)
+	}
+	mw.ss.span.SetAttributes(attributes...)
+	return nil
+}
+
+// Error records the given error against the span and sets the span to an error status
+func (h *Hook) Error(hookContext of.HookContext, err error, hookHints of.HookHints) {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if ok {
+		mw.ss.span.RecordError(err)
+		mw.ss.span.SetStatus(codes.Error, err.Error())
+	}
+}
+
+// Finally this method is unused for this hook, spans are closed via context
+func (h Hook) Finally(hookContext of.HookContext, hookHints of.HookHints) {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if ok {
+		mw.ss.span.End()
+		mw.ss.cancel()
```

Comment:
What does cancel do?

Reply:
cancel closes the context linked to this request and unblocks the `goroutine` started in the `Before` method. The same happens if the parent context is closed.